### PR TITLE
Bring iris-grib back into iris

### PR DESCRIPTION
**Pull request by marqh**
* [x] functional code transferred
* [x] unit tests transferred
* [ ] ~~outstanding test issues skipped~~

### CODE REVIEW

File: lib/iris/fileformats/__init__.py [hunk 1]
```diff
@@ -27,13 +27,11 @@
...
+
 try:
-    import iris_grib as igrib
+    from . import grib as igrib
 except ImportError:
-    try:
-        from . import grib as igrib
-    except ImportError:
-        igrib = None
+    igrib = None
```
Comment (lbdreyer) [inline lib/iris/fileformats/__init__.py]: Why is this necessary? Why not just:
`from . import grib as igrib`

File: lib/iris/fileformats/grib/__init__.py [hunk 2]
```diff
@@ -41,56 +35,44 @@
...
+from . import _save_rules
+from ._load_convert import convert as load_convert
+from .message import GribMessage
 
-# Issue a blanket deprecation for this module.
-warn_deprecated(
-    "The module iris.fileformats.grib is deprecated since v1.10. "
-    "Please install the package 'iris_grib' package instead.")
 
+__version__ = '0.10.0-dev'
```
Comment (lbdreyer) [inline lib/iris/fileformats/grib/__init__.py]: This will need to be removed.

File: lib/iris/fileformats/__init__.py (hunk 1 above)
Comment (marqh) [inline lib/iris/fileformats/__init__.py]: at present the grib module is dependent on gribapi which may not be available, so a conditional import is used
e.g. windows, python3.x

adopting ECCodes may help this, but we may still maintain a conditional import and keep GRIB as an optional dependency

either way, for now it is required

**Comment by marqh** [comment]
I believe this PR now represents a functional, backwards incompatible change

File: lib/iris/fileformats/__init__.py (hunk 1 above)
Comment (lbdreyer) [inline lib/iris/fileformats/__init__.py]: Is there a specific reason you've chosen to import it as `igrib`? It used to be imported as just `grib`

File: lib/iris/tests/unit/fileformats/grib/test_save_messages.py [hunk 3]
```diff
@@ -25,10 +25,10 @@
 import iris.tests as tests
 
 import gribapi
+import mock
```
Comment (lbdreyer) [inline lib/iris/tests/unit/fileformats/grib/test_save_messages.py]: Perhaps all these should go back to using the mock from iris.tests, i.e.
`from iris.tests import mock`

File: lib/iris/tests/unit/fileformats/grib/test_save_messages.py [hunk 4]
```diff
@@ -1,4 +1,4 @@
-# (C) British Crown Copyright 2015, Met Office
+# (C) British Crown Copyright 2016 - 2017, Met Office
```
Comment (lbdreyer) [inline lib/iris/tests/unit/fileformats/grib/test_save_messages.py]: 2016-2017 may have been correct for iris-grib but this file existed in Iris from 2015

File: lib/iris/tests/unit/fileformats/grib/save_rules/test_product_definition_template_40.py [hunk 5]
```diff
@@ -52,10 +53,10 @@ def test_constituent_type(self, mock_set):
...
         mock_set.assert_any_call(mock.sentinel.grib,
-                                 "productDefinitionTemplateNumber", 40)
+                                 'productDefinitionTemplateNumber', 40)
         mock_set.assert_any_call(mock.sentinel.grib,
-                                 "constituentType", 0)
+                                 'constituentType', 0)
 
 
-if __name__ == "__main__":
+if __name__ == '__main__':
```
Comment (lbdreyer) [inline lib/iris/tests/unit/fileformats/grib/save_rules/test_product_definition_template_40.py]: I have notice that some test uses single quotes(`'string'`) whereas others use double quotes (`"string"`).

Of course this shouldn't be dealt with in this PR, but it's interesting this file is specifically changed from double quotes to single

File: lib/iris/tests/unit/fileformats/grib/save_rules/test__missing_forecast_period.py [hunk 6]
```diff
@@ -28,9 +28,11 @@
 import iris.tests as tests
 
 import datetime
+import numpy as np
```
Comment (lbdreyer) [inline lib/iris/tests/unit/fileformats/grib/save_rules/test__missing_forecast_period.py]: numpy doesn't actually get used

File: lib/iris/tests/unit/fileformats/grib/message/test__MessageLocation.py [hunk 7]
```diff
@@ -15,7 +15,7 @@
 # You should have received a copy of the GNU Lesser General Public License
 # along with Iris.  If not, see <http://www.gnu.org/licenses/>.
 """
-Unit tests for the `iris.fileformats.grib.message._MessageLocation` class.
+Unit tests for the `iris.message._MessageLocation` class.
```
Comment (lbdreyer) [inline lib/iris/tests/unit/fileformats/grib/message/test__MessageLocation.py]: This must be wrong. I can't find anything about iris.message

File: lib/iris/tests/unit/fileformats/grib/message/test__DataProxy.py [hunk 8]
```diff
@@ -15,7 +15,7 @@
 # You should have received a copy of the GNU Lesser General Public License
 # along with Iris.  If not, see <http://www.gnu.org/licenses/>.
 """
-Unit tests for the `iris.fileformats.grib.message._DataProxy` class.
+Unit tests for the `iris.message._DataProxy` class.
```
Comment (lbdreyer) [inline lib/iris/tests/unit/fileformats/grib/message/test__DataProxy.py]: This also needs to be changed back.

File: lib/iris/tests/unit/fileformats/grib/message/test_GribMessage.py [hunk 9]
```diff
@@ -252,14 +248,12 @@ def section_3(self, scanning_mode):
...
 
 
-@tests.iristest_timing_decorator
 class Test_data__grid_template_40_regular(tests.IrisTest_nometa,
                                           Mixin_data__grid_template):
     def section_3(self, scanning_mode):
         return _example_section_3(40, scanning_mode)
 
 
-@tests.iristest_timing_decorator
```
Comment (lbdreyer) [inline lib/iris/tests/unit/fileformats/grib/message/test_GribMessage.py]: Did you want to reinstate these?

File: lib/iris/tests/unit/fileformats/grib/load_convert/test_time_range_unit.py [hunk 10]
```diff
@@ -22,13 +22,13 @@
...
-# import iris tests first so that some things can be initialised
+# import iris.tests first so that some things can be initialised
 # before importing anything else.
 import iris.tests as tests
 
 from cf_units import Unit
-
 from iris.exceptions import TranslationError
+
 from iris.fileformats.grib._load_convert import time_range_unit
```
Comment (lbdreyer) [inline lib/iris/tests/unit/fileformats/grib/load_convert/test_time_range_unit.py]: Admittedly I had been ignoring these extra lines, but I actually they should be removed.

File: lib/iris/tests/unit/fileformats/grib/load_convert/test_reference_time_coord.py [hunk 11]
```diff
@@ -61,6 +62,11 @@ def _check(self, section, standard_name=None):
 
     def test_start_of_forecast(self):
         section = deepcopy(self.section)
+        section['significanceOfReferenceTime'] = 0
+        self._check(section, 'forecast_reference_time')
+
+    def test_start_of_forecast(self):
```
Comment (lbdreyer) [inline lib/iris/tests/unit/fileformats/grib/load_convert/test_reference_time_coord.py]: These tests have the same name , surely that must cause problems!

File: lib/iris/tests/unit/fileformats/grib/load_convert/test_product_definition_template_31.py [hunk 12]
```diff
@@ -22,25 +22,23 @@
...
+# import iris.tests first so that some things can be initialised
 # before importing anything else.
 import iris.tests as tests
 
 from copy import deepcopy
-import warnings
-
+import mock
 import numpy as np
+import warnings
```
Comment (lbdreyer) [inline lib/iris/tests/unit/fileformats/grib/load_convert/test_product_definition_template_31.py]: Surely `warnings` is a builtin and therefore should be in the previous group of imports?

File: lib/iris/tests/unit/fileformats/grib/load_convert/test_product_definition_template_0.py [hunk 13]
```diff
@@ -23,15 +23,18 @@
...
-# import iris tests first so that some things can be initialised
+# import iris.tests first so that some things can be initialised
 # before importing anything else.
 import iris.tests as tests
 
+import mock
+
 import iris.coords
+
+import iris.fileformats.grib
```
Comment (lbdreyer) [inline lib/iris/tests/unit/fileformats/grib/load_convert/test_product_definition_template_0.py]: Why is this import necessary?

File: lib/iris/tests/unit/fileformats/grib/load_convert/test_grid_definition_template_4_and_5.py [hunk 14]
```diff
@@ -23,42 +23,36 @@
...
+# import iris.tests first so that some things can be initialised
 # before importing anything else.
 import iris.tests as tests
 
 from copy import deepcopy
-import warnings
-
+import mock
 import numpy as np
+import warnings
```
Comment (lbdreyer) [inline lib/iris/tests/unit/fileformats/grib/load_convert/test_grid_definition_template_4_and_5.py]: Again, surely `warnings` is a standard library?

File: lib/iris/tests/unit/fileformats/grib/load_convert/test_grib2_convert.py [hunk 15]
```diff
@@ -19,15 +19,15 @@
...
-# import iris tests first so that some things can be initialised
+# import iris.tests first so that some things can be initialised
 # before importing anything else.
 import iris.tests as tests
 
 import copy
+import mock
 
-import iris
+import iris.fileformats.grib
```
Comment (lbdreyer) [inline lib/iris/tests/unit/fileformats/grib/load_convert/test_grib2_convert.py]: Again, why is this imported?

File: lib/iris/tests/unit/fileformats/grib/grib_phenom_translation/test_grib_phenom_translation.py [hunk 16]
```diff
@@ -0,0 +1,168 @@
...
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU Lesser General Public License for more details.
+#
+# You should have received a copy of the GNU Lesser General Public License
+# along with Iris.  If not, see <http://www.gnu.org/licenses/>.
+'''
+Unit tests for the mod:`iris.fileformats.grib.grib_phenom_translation` module.
+
+Carried over from old iris/tests/test_grib_phenom_translation.py.
+Code is out of step with current test conventions and standards.
```
Comment (lbdreyer) [inline lib/iris/tests/unit/fileformats/grib/grib_phenom_translation/test_grib_phenom_translation.py]: This does look almost identical to iris/tests/test_grib_phenom_translation.py, so either this or that file should be removed (I vote that file).

Comment (lbdreyer) [inline lib/iris/tests/unit/fileformats/grib/grib_phenom_translation/test_grib_phenom_translation.py]: If this file gets deleted, then  lib/iris/tests/unit/fileformats/grib/grib_phenom_translation/__init__.py  will also need deleting.

File: lib/iris/tests/test_coding_standards.py [hunk 17]
```diff
@@ -85,9 +85,8 @@ class StandardReportWithExclusions(pep8.StandardReport):
         '*/iris/analysis/_interpolate_private.py',
         '*/iris/fileformats/cf.py',
         '*/iris/fileformats/dot.py',
-        '*/iris/fileformats/grib/__init__.py',
```
Comment (lbdreyer) [inline lib/iris/tests/test_coding_standards.py]: Why have you removed the __init__?

File: lib/iris/fileformats/grib/_load_convert.py [hunk 18]
```diff
@@ -38,10 +38,13 @@
 import iris.coord_systems as icoord_systems
 from iris.coords import AuxCoord, DimCoord, CellMethod
 from iris.exceptions import TranslationError
-from iris.fileformats.grib import grib_phenom_translation as itranslation
+from . import grib_phenom_translation as itranslation
 from iris.fileformats.rules import ConversionMetadata, Factory, Reference
 from iris.util import _is_circular
 
+from ._grib1_load_rules import grib1_convert
+from .message import GribMessage
```
Comment (lbdreyer) [inline lib/iris/fileformats/grib/_load_convert.py]: Theses imports should be moved.

File: lib/iris/fileformats/grib/_grib_cf_map.py [hunk 19]
```diff
@@ -1,26 +1,26 @@
...
 #
-# Iris is distributed in the hope that it will be useful,
+# iris is distributed in the hope that it will be useful,
 # but WITHOUT ANY WARRANTY; without even the implied warranty of
 # MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 # GNU Lesser General Public License for more details.
 #
 # You should have received a copy of the GNU Lesser General Public License
-# along with Iris.  If not, see <http://www.gnu.org/licenses/>.
+# along with iris.  If not, see <http://www.gnu.org/licenses/>.
```
Comment (lbdreyer) [inline lib/iris/fileformats/grib/_grib_cf_map.py]: These should be `Iris` with a capital `I` to be consistent with the other files.

**Comment by pelson** [comment]
Hi @marqh. Has this been discussed somewhere in the public domain? I'm personally hugely :-1: on doing this. I'd like to have a suitable place to have that discussion (a PR isn't the right place).

**Comment by pelson** [comment]
On that front, I keep getting caught out by feature branch merge requests. Has feature branch usage been discussed anywhere? It isn't clear to me *when* we should be reviewing this stuff when it comes to feature branches. Again, a conversation to be had outside of this PR. Ping @bjlittle.

File: lib/iris/tests/unit/fileformats/grib/load_convert/test_product_definition_template_0.py (hunk 13 above)
Comment (marqh) [inline lib/iris/tests/unit/fileformats/grib/load_convert/test_product_definition_template_0.py]: it is used at line 93
```
        convert_options = iris.fileformats.grib._load_convert.options
```

File: lib/iris/tests/unit/fileformats/grib/load_convert/test_grib2_convert.py (hunk 15 above)
Comment (marqh) [inline lib/iris/tests/unit/fileformats/grib/load_convert/test_grib2_convert.py]: it is used at line 64
```
        this = iris.fileformats.grib._load_convert
```

File: lib/iris/tests/test_coding_standards.py (hunk 17 above)
Comment (marqh) [inline lib/iris/tests/test_coding_standards.py]: because it passes the coding standards tests now, so doesn't need to be skipped

File: lib/iris/fileformats/grib/_load_convert.py (hunk 18 above)
Comment (marqh) [inline lib/iris/fileformats/grib/_load_convert.py]: moved to where please?

File: lib/iris/tests/unit/fileformats/grib/message/test_GribMessage.py (hunk 9 above)
Comment (marqh) [inline lib/iris/tests/unit/fileformats/grib/message/test_GribMessage.py]: yes, done

**Comment by marqh** [comment]
@pelson 

> Has this been discussed somewhere in the public domain? I'm personally hugely :-1: on doing this. I'd like to have a suitable place to have that discussion (a PR isn't the right place).

I had thought that @bjlittle had started a thread on the forum about this, as part of the development team developing a plan for this work late last week
https://github.com/SciTools/iris/projects/4
but it seems that step was missed, so I have made a discussion thread
https://groups.google.com/forum/#!topic/scitools-iris-dev/lFw2_CnaEHU
for this purpose

This PR is part of the feasibility study, so I would like to keep it focused on that, and not have wider concerns blocking any progress

This seems consistent with your comments

File: lib/iris/tests/unit/fileformats/grib/test_save_messages.py [hunk 20]
```diff
@@ -25,10 +25,10 @@
 import iris.tests as tests
 
 import gribapi
+from iris.tests import mock
```
Comment (lbdreyer) [inline lib/iris/tests/unit/fileformats/grib/test_save_messages.py]: @marqh This mock import is in the wrong location (and this is also the case for a few other files).

**Comment by marqh** [comment]
@lbdreyer I think these changes cover all of the points you have raised so far.

this is ready for a second pass, I believe

File: lib/iris/fileformats/grib/__init__.py [hunk 21]
```diff
@@ -237,102 +178,72 @@ def __init__(self, grib_message, grib_fh=None, auto_regularise=True):
...
     def _confirm_in_scope(self):
         """Ensure we have a grib flavour that we choose to support."""
 
-        #forbid alternate row scanning
-        #(uncommon entry from GRIB2 flag table 3.4, also in GRIB1)
+        # forbid alternate row scanning
+        # (uncommon entry from GRIB2 flag table 3.4, also in GRIB1)
         if self.alternativeRowScanning == 1:
-            raise iris.exceptions.IrisError("alternativeRowScanning == 1 not handled.")
+            raise ValueError("alternativeRowScanning == 1 not handled.")
```
Comment (lbdreyer) [inline lib/iris/fileformats/grib/__init__.py]: Should we revert this to `iris.exceptions.IrisError`

File: lib/iris/fileformats/grib/__init__.py [hunk 22]
```diff
@@ -1053,38 +828,24 @@ def save_pairs_from_cube(cube):
         yield (slice2D, grib_message)
 
 
-def as_messages(cube):
-    """
-    .. deprecated:: 1.10
```
Comment (lbdreyer) [inline lib/iris/fileformats/grib/__init__.py]: What is the process now that `as_messages` and `as_pairs` were deprecated in Iris 1.10. Are we ready to completely remove them from the codebase?

File: lib/iris/fileformats/grib/_save_rules.py [hunk 23]
```diff
@@ -573,14 +572,6 @@ def _non_missing_forecast_period(cube):
                       "scaling required.")
     fp = int(fp)
 
-    # Turn negative forecast times into grib negative numbers?
-    from iris.fileformats.grib import hindcast_workaround
-    if hindcast_workaround and fp < 0:
-        msg = "Encoding negative forecast period from {} to ".format(fp)
-        fp = 2**31 + abs(fp)
-        msg += "{}".format(np.int32(fp))
-        warnings.warn(msg)
```
Comment (lbdreyer) [inline lib/iris/fileformats/grib/_save_rules.py]: Should this be deleted, the hindcast_workaround looks like it was only introduced recently

Comment (marqh) [inline lib/iris/fileformats/grib/_save_rules.py]: i don't think this is the case, i think this is a legacy implementation which has no been overtaken by new templates, and hence removed totally from iris-grib

File: lib/iris/fileformats/grib/__init__.py (hunk 22 above)
Comment (marqh) [inline lib/iris/fileformats/grib/__init__.py]: yes, ready to remove;

these were misnamed and then renamed, hence the deprecation; they're part of the v2 removal

**Comment by marqh** [comment]
Hi @lbdreyer 

I have made all the changes we discussed, and commented on the two which I have not followed up on

I have also 'squished' the review commits into one

I think the commit history is now concise and useful.  If you are happy to merge this, please implement a merge commit, do not squash and merge, to preserve the key change steps

thank you
mark

File: lib/iris/fileformats/grib/_save_rules.py (hunk 23 above)
Comment (lbdreyer) [inline lib/iris/fileformats/grib/_save_rules.py]: Yep sorry, thought it was added in Iris **1.12** but actually it was added in Iris **1.2**

File: lib/iris/tests/unit/fileformats/grib/save_rules/test_grid_definition_template_12.py [hunk 24]
```diff
@@ -31,11 +31,10 @@
 
 import iris.coords
 from iris.coord_systems import GeogCS, TransverseMercator
+from iris.fileformats.grib._save_rules import grid_definition_template_12
 from iris.exceptions import TranslationError
```
Comment (lbdreyer) [inline lib/iris/tests/unit/fileformats/grib/save_rules/test_grid_definition_template_12.py]: **Very minor** but these imports aren't in alphabetical order.

**Comment by lbdreyer** [comment]
Thanks @marqh !

@pelson Our next steps will be to daskify the grib loading/saving and then merge this `gribiris` feature branch onto the `dask` feature branch. We would like to do this in the next few days so please raise your concerns, if you have them, on the [google groups post](https://groups.google.com/forum/#!topic/scitools-iris-dev/lFw2_CnaEHU) that @marqh created, as soon as is feasible.
